docs/source/column-metadata.rst

```diff
@@ -0,0 +1,185 @@
+
+   .. code-block:: sh
+
+      q) select t_id.name from s
+      name
+      ----
+      b
+      c
+      a
+      b
+      c
+
+
+Other Metadata
+==============
+
+Additionally, KDB+ exposes metadata regarding the structure of a particular column. In particular it allows one to attribute different properties to a column. Those properties are:
+
+Sorted
+------
+
+This indicates that a column in sorted in ascending order. Lookups will be done with binary search rather than a linear scan of the column and operations that can take advantage of sortedness (e.g., ``min`` and ``max``) will be greatly sped up.
+
+Unique
+------
+
+Uniqueness indicates that the values are in a column are distinct. Certain operations such as ``distinct`` and ``nunique`` are sped up many times.
+
+Grouped
+-------
+
+According to the ``q`` documentation the Grouped attributed roughly corresponds to a SQL index. It is implemented by using a hash table that maps unique values to a list of positions. This can speed up queries with a ``where`` clauses by allowing the query engine to immediately access the positions the query should run over.
+
+Example:
+
+.. code-block:: python
+
+   >>> [0, 1, 2, 0, 1, 2, 1, 2, 0, 0, 1, 2, 0]  # lots of repetition, but not really any runs
+
+Parted
+------
+
+This attribute indicates that there are runs of the same value in our data. In this case ``q`` will store a hash table mapping unique values to the start position of each run. It likely stores the length of the run as well.
+
+Example:
+
+.. code-block:: python
+
+   >>> [1, 1, -1, -1, -1, 2, 2, 2, 3]  # lots of repetition with runs
+
+Implementation
+==============
+
+Regardless of how these properties change computation, ``blaze`` needs to be able to tell other systems about metadata where possible. This is one of the main benefits to using a relational database, namely, the ability to relate information.
+
+Sorted
+------
+
+TODO
+
+Unique and Grouped Attributes
+-----------------------------
+
+Categorical Types
+^^^^^^^^^^^^^^^^^
+Categorical types can cover the cases of unique and grouped. We implement this in the ``datashape`` library. It has two relevant properties:
+
+1. A list of unique sorted categories.
+2. A list of the indices of the categories from item 1 as unsigned integers of the minimum type necessary to support the number of unique categories.
+
+Primary and Foreign Keys
+------------------------
+
+It is important to note that we want to keep foreign key metadata separate from type information. Adding foreign key metadata to a blaze expression will thus require an extra step.
+
+   .. code-block:: python
+
+      >>> t = symbol('t', """var * {
+      ...     id: int64,
+      ...     name: categorical[type=string, categories=["Alice", "Bob", "Joel"]],
+      ...     amount: float64
+      ... }""")
+      >>> t.primary_key = t.id
+      >>> s = symbol('s', "var * {t_id: int64, height: float64}")
+      >>> s.relations = {s.t_id: t.id}
```

Comment:
I left this in there to show the evolution of my thoughts, but now that it's in the commit history I'll blow it away. Hooray git!